Document design of tier 2 engines (#640)
faster-cpython/ideas: 3.13/engine.md

executor that can also exit before it does any work, or it exits to itself,
we can find ourselves in an infinite loop.

There are two possible approaches to avoiding this:
1. We track which executors might not make
progress and are careful about which executors we linked together, or
2. We require all executors to make progress.

1 has many edge cases which makes it hard to reason about and get correct.

2 is simpler, but may be a bit slower as we may need to pessimize the first instruction.

We will use approach 2. It is much easier to reason about and should be almost as fast.

If all executors are guaranteed to make progress, then transfers from one executor 
to another can be implemented by a single jump/tail-call.

### Inter-superblock optimization

Some superblocks can be quite short, but form a larger region of hot code, that 
we want to optimize. In order to do that we want to propogate type information
and representation changes across edges, which means storing that information
on exits when they are cold. Since most exits will remain cold, we need to 
store this information in a compact form.

### Making "hot" exits fast and "cold" exits small

> In order to make hot exits fast, we them to be as simple as possible,

Reviewer: ```suggestion
In order to make hot exits fast, we need them to be as simple as possible,
```

passing as little information as possible.

We also want to make cold exits as small as possible, but cold exits
may become hot exits, so we need them both to use the same interface.

## The implementation

This section describes one possible implemnentation. The initial implementation
will not support inter-superblock optimization, but we should plan to support it
in the future.

### Making "hot" exits fast

In order to make hot exits fast, we want to implement them as a single, unconditional jump.
In the tier 2 interpreter, that will be implemented by setting the IP to the first

> instruction of the target executor.

Reviewer 2: Isn't this more specifically also decrefing the old executor, increfing the new executor etc as described below?  (And I think also updating a "current executor" pointer, correct?)

Maybe it's not relevant to repeat here, but generally speaking, it's "moving the IP + some bookkeeping".  Or is the bookkeeping not necessary when hot for some non-obvious reason?

required once the exit becomes hot, so we want to store that information in a way
that minimizes the cost to hot exits and minimizes the memory used for cold exits.
That information is:
* The offset/location of the pointer to the exit in the executor, so it can be updated.
* The target (offset into the code object of the tier 1 instruction)
* Any relevant known type information (this is optional but will improve optimization)
* Any representation changes that have been made.
* The "hotness" counter

### Minimizing memory use

There are number of ways we can reduce memory use, for example:
* Putting things in (ideally const) arrays, so we can refer to them by a one or two byte index,
  instead of an eight byte pointers.
* For complex data like type information and representation changes, use trees or deltas,
  so that the information like `(A, B, C)` can be stored as `(A, B) <- C` allowing `(A, B)` to
  be shared.

### Each executor gets a table of exit data

Each executor will get a table of exit data. We can compute the size of this
when creating the executor. The entries in this table are described below.

### Fixed number of exit objects

> Since the vast majority of exits will not need patching, we do not want to pass the offset,

Author: Yes, replacing the exit with an executor.

> Since the vast majority of exits will not need patching, we do not want to pass the offset,

Author: The offset of the exit in the executor, so that the exit knows where to update the pointer

so we need to store the offset in the executor.

Since there is a fixed number of micro-ops allowed in a superblock, we have an uppper

> bound on the offset. We will-preallocate one exit object per possible offset.

Reviewer: ```suggestion
bound on the offset. We will preallocate one exit object per possible offset.
```

### Exit data

* The offset/location of the pointer: Implicit if we embed the exit pointer in the exit data
* The target: Has a maximum value of about 10**9, so store as a `uint32_t`

> * Any relevant known type information: Format TBD.

Reviewer 2: Am I correct that this is only needed for intra-executor optimization and thus could be deferred until we start working on that?

Author: yes

  We don't need to decide on the format of this data for now.
* Any representation changes that have been made: Likewise, format TDB.
* The "hotness" counter. See below.

#### Representation changes

Representation changes are things like storing the top value(s) of the stack in registers, or performing scalar replacement on objects or frames.
In order to drop back to tier1, we need to be able to undo those changes, so we need to record them.

#### Hotness counters

We need to track how hot an exit is. Ideally we want to be able to distinguish
between an exit that is hot, and one that is cold but has exited a large number of
times over prolonged execution.

There are three ways we can store counters:
1. Store a counter for each exit in the superblock.

> 2. Have one exit per counter, and change the exit object to change the counter.

Author: Changed to "Have one exit per possible value of the counter"

This means that the micro ops for the first tier 1 instruction in an executor
cannot contain an `EXIT_IF`. In practice this means that before translating
to micro-ops, the first instruction must be coverted to its unspecialized form.

### Exiting to invalid executors

We want `EXIT_IF` to be efficient, so it needs to enter the next executor unconditionally.
This means that when we invalidate an executor, we need to modify that executor so that
it immediately drops into tier 1 upon being executed. Since tier 1 will never enter an invalid
executor, we are thus guaranteed not to execute an invalid executor.
In the tier 2 interpreter, we change the first instruction to `EXIT_TRACE`.
In the JIT, we will need to change the function pointer to point to a function that returns to tier 1.

### The mechanics of transferring execution between executors

When transfering control we need to:
1. Incref the new executor
2. Set `current_executor` to the new executor 
3. Decref the old executor.

In the future, when we have deferred references, we can make the current executor a deferred reference
and skip the incref/decref.

#### JIT compiler

> To tranfer control between executors, we make a jump, implemented as an indirect

Reviewer: ```suggestion
To transfer control between executors, we make a jump, implemented as an indirect
```

 tail-call in the
generated stencil.

The generated code must decref the old executor, as we cannot decref the old executor when still executing it, meaning we must pass the old executor as an argument in the tail call.

#### Interpreter 

> We can do the three steps (incref, update current exexutor, decref) before entering the new executor, since the we don't need to worrying about freeing code that we are running.

Reviewer: ```suggestion
We can do the three steps (incref, update current executor, decref) before entering the new executor, since then we don't need to worry about freeing code that we are running.
```
One more typo
